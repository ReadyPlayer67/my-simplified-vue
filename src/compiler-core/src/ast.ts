export const enum NodeTypes {
    INTERPOLATION,//插值
    SIMPLE_EXPRESSION,//简单表达式
}