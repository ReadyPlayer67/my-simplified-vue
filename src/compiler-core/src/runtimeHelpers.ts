export const TO_DISPLAY_STRING = Symbol('toDisplayString')

export const helperNameMap = {
    [TO_DISPLAY_STRING]: 'toDisplayString'
}