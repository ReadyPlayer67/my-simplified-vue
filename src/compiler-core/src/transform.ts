export function transform(root, options = {}) {
    //创建一个上下文存放root和options
    const context = createTransformContext(root, options)
    traverseNode(root, context)
    //创建一个codegenNode作为生成render函数的入口节点
    createCodegenNode(root)
}

function createCodegenNode(root){
    root.codegenNode = root.children[0]
}

function createTransformContext(root, options) {
    return {
        root,
        nodeTransforms: options.nodeTransforms || []
    }
}

//利用递归对ast树进行深度优先遍历
function traverseNode(node, context) {
    const {nodeTransforms} = context
    //通过插件机制将容易变动的代码抽离出去，由外部去实现
    //这样程序的扩展性就变得很强了，并且提高了程序的可测试性
    for (const transform of nodeTransforms) {
        transform(node)
    }
    traverseChildren(node, context)
}

function traverseChildren(node, context) {
    const children = node.children
    if (children) {
        for (let i = 0; i < children.length; i++) {
            traverseNode(children[i], context)
        }
    }
}
