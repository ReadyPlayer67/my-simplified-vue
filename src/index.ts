//mini-vue出口
export * from './runtime-dom'