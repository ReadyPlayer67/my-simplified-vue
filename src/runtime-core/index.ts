export {createApp} from './createApp'
//导出h方法
export {h} from './h'