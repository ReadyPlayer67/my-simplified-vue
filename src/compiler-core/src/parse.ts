import {NodeTypes} from "./ast";

const enum TagType {
    Start,
    End
}

//得到抽象语法树AST
export function baseParse(content: string) {
    const context = createParserContext(content)
    return createRoot(parseChildren(context, ''))
}

function parseChildren(context, parentTag) {
    let nodes: any = []
    while (!isEnd(context, parentTag)) {
        let node
        const s = context.source
        if (s.startsWith('{{')) {
            node = parseInterpolation(context)
        } else if (s[0] === '<') {//如果是<开头，认为是一个element标签
            if (/[a-z]/i.test(s[1])) {//<后面必须是字母
                node = parseElement(context)
            }
        }
        if (!node) {
            node = parseText(context)
        }
        nodes.push(node)
    }
    return nodes
}

function isEnd(context, parentTag) {
    const s = context.source
    //当source没有值或者遇到结束标签的时候，应当停止循环
    //创建root AST时parentTag传的空，如果是空就不需要对闭合标签进行判断了
    if (parentTag && s.startsWith(`</${parentTag}>`)) {
        return true
    }
    return !s
}

function parseText(context) {
    const endToken = '{{'
    let endIndex = context.source.length
    const index = context.source.indexOf(endToken)
    if (index !== -1) {
        endIndex = index
    }
    const content = parseTextData(context, endIndex)
    return {
        type: NodeTypes.TEXT,
        content
    }
}

function parseTextData(context, length) {
    const content = context.source.slice(0, length)
    advanceBy(context, length)
    return content
}

function parseElement(context) {
    //处理<div>
    const element: any = parseTag(context, TagType.Start)
    //在开始标签和闭合标签中间用parseChildren处理子节点内容
    //把element标签名字传进去，这样在isEnd控制循环停止时就可以拿到闭合标签名
    element.children = parseChildren(context, element.tag)
    //处理</div>闭合标签，保证推进
    parseTag(context, TagType.End)
    return element
}

function parseTag(context, type: TagType) {
    //利用()实现分组捕获，这样在match[1]就能拿到tag内容
    const match: any = /^<\/?([a-z]*)/i.exec(context.source)
    // console.log('match:', match)
    //match[1]就是匹配到的tag类型
    const tag = match[1]
    //match[0]是匹配的全部字符串，也就是<div，往前推进<div和>的字符长度
    advanceBy(context, match[0].length)
    advanceBy(context, 1)
    //用一个标记代表处理的是<div>还是</div>闭合标签，如果是闭合标签就不用返回ast节点
    if (type === TagType.End) return
    return {
        type: NodeTypes.ELEMENT,
        tag
    }
}

function parseInterpolation(context) {
    //定义前后两个分隔符
    const openDelimiter = '{{'
    const closeDelimiter = '}}'
    //查找闭合位置
    const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)
    //往前推进两个字符，去除掉'{{'
    advanceBy(context, openDelimiter.length)
    //通过closeIndex减掉'{{'的长度（因为已经推进了2个字符）获得content的长度
    const rawContentLength = closeIndex - openDelimiter.length
    //根据content长度截取context.source获取content
    const rawContent = parseTextData(context, rawContentLength)
    const content = rawContent.trim()
    //把已经处理过的内容删掉，继续往前推进处理后面的内容
    advanceBy(context, closeDelimiter.length)
    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content
        }
    }
}

function advanceBy(context, length) {
    context.source = context.source.slice(length)
}

function createRoot(children) {
    return {
        children
    }
}

function createParserContext(content: string) {
    return {
        source: content
    }
}