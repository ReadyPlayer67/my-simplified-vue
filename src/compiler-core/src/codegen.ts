import {NodeTypes} from "./ast";
import {CREATE_ELEMENT_VNODE, helperNameMap, TO_DISPLAY_STRING} from "./runtimeHelpers";

export function generate(ast) {
    const context = createCodegenContext()
    const {push} = context
    genFunctionPreamble(ast, context)
    const functionName = 'render'
    const args = ['_ctx', '_cache']
    const signature = args.join(', ')
    push(`function ${functionName} (${signature}) {`)
    push('return ')
    genNode(ast.codegenNode, context)
    push('}')
    return {
        code: context.code
    }
}

function createCodegenContext() {
    const context = {
        code: '',
        push(source) {
            context.code += source
        },
        helper(key) {
            return `_${helperNameMap[key]}`
        }
    }
    return context
}

//处理前导码（import，const）这些
function genFunctionPreamble(ast, context) {
    const {push} = context
    const VueBinging = 'Vue'
    const aliasHelper = (s) => `${helperNameMap[s]}: _${helperNameMap[s]}`
    if (ast.helpers.length > 0) {
        push(`const { ${ast.helpers.map(helper => aliasHelper(helper)).join(', ')} } = "${VueBinging}"`)
    }
    push('\n')
    push('return ')
}

function genNode(node, context) {
    switch (node.type) {
        case NodeTypes.INTERPOLATION:
            genInterpolation(node, context)
            break;
        case NodeTypes.TEXT:
            genText(node, context);
            break;
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(node, context)
            break;
        case NodeTypes.ELEMENT:
            genElement(node, context)
            break;
        default:
            break;
    }

}

function genText(node, context) {
    const {push} = context
    push(`'${node.content}'`)
}

function genInterpolation(node, context) {
    const {push, helper} = context
    push(`${helper(TO_DISPLAY_STRING)}(`)
    genNode(node.content, context)
    push(')')
}

function genExpression(node, context) {
    const {push} = context
    push(`${node.content}`)
}

function genElement(node, context) {
    const {push, helper} = context
    const {tag} = node
    push(`${helper(CREATE_ELEMENT_VNODE)}(`)
    push(`"${tag}")`)
}