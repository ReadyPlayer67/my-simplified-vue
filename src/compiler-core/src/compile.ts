import {baseParse} from "./parse";
import {transform} from "./transform";
import {transformExpression} from "./transforms/transformExpression";
import {transformElement} from "./transforms/transformElement";
import {transformText} from "./transforms/transformText";
import {generate} from "./codegen";

//暴露一个出口方法baseCompile
export function baseCompile(template: string) {
    const ast: any = baseParse(template)
    transform(ast, {
        nodeTransforms: [transformExpression, transformElement, transformText]
    })
    return generate(ast)
}