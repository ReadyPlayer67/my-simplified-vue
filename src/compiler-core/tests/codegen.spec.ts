import {baseParse} from "../src/parse";
import {generate} from "../src/codegen";

describe('codegen',function (){
    it('string', function () {
        const ast = baseParse('hi')
        const code = generate(ast)
        //生成一个快照,之后每次测试生成结果和快照进行对比
        expect(code).toMatchSnapshot()
    });
})
