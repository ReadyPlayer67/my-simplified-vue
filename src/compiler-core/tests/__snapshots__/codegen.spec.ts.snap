// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
Object {
  "code": "const { createElementVNode: _createElementVNode } = \\"Vue\\"
return function render (_ctx, _cache) {return _createElementVNode(\\"div\\" ,null, )}",
}
`;

exports[`codegen interpolation 1`] = `
Object {
  "code": "const { toDisplayString: _toDisplayString } = \\"Vue\\"
return function render (_ctx, _cache) {return _toDisplayString(_ctx.message)}",
}
`;

exports[`codegen string 1`] = `
Object {
  "code": "
return function render (_ctx, _cache) {return 'hi'}",
}
`;
